x-pack/plugins/features/server/feature_registry.ts

```diff
@@ -25,28 +25,31 @@ export class FeatureRegistry {
       throw new Error(`Feature with id ${feature.id} is already registered.`);
     }
 
-    const featureCopy: Feature = cloneDeep(feature as Feature);
+    const featureCopy: FeatureConfig = cloneDeep(feature as FeatureConfig);
 
-    this.features[feature.id] = applyAutomaticPrivilegeGrants(featureCopy as Feature);
+    this.features[feature.id] = applyAutomaticPrivilegeGrants(featureCopy as FeatureConfig);
   }
 
   public getAll(): Feature[] {
     this.locked = true;
-    return cloneDeep(Object.values(this.features));
+    return Object.values(this.features).map(featureConfig => new Feature(featureConfig));
```

Comment:
I think we _could_. The previous implementation was dealing with simple objects, rather than class instances, and each call to `getAll` would return a deeply cloned array of feature definitions. The idea of fresh instances here was to maintain parity with the previous implementation. The goal is to prevent consumers from modifying a retrieved feature definition, and having that modification impact other consumers.

Unless you have a strong objection, I'd like to leave this as-is for the time being. The set of features is pretty small, and there are only a couple of places where `getAll` is called, so I don't think the overhead here is something to be concerned about yet.

Reply:
no objection, the overhead is insignificant here.